riscv/src/register/macros.rs

```diff
@@ -397,3 +397,446 @@ macro_rules! clear_pmp {
+            pub const BITMASK: usize = $mask;
+
+            /// Creates a new CSR type from raw bits value.
+            ///
+            /// Only bits in the [BITMASK](Self::BITMASK) will be set.
+            pub const fn from_bits(bits: usize) -> Self {
+                Self { bits: bits & $mask }
+            }
+
+            /// Gets the raw bits value.
+            pub const fn bits(&self) -> usize {
+                self.bits & $mask
+            }
+
+            /// Gets the bitmask for the CSR type's bitfields.
+            pub const fn bitmask(&self) -> usize {
+                Self::BITMASK
+            }
+        }
+
+        $crate::read_csr_as!($ty, $csr);
+    };
+}
+
+/// Helper macro to create a read-only CSR type.
+///
+/// The type allows to read the CSR value into memory.
+#[macro_export]
+macro_rules! write_only_csr {
+    ($doc:expr,
+     $ty:ident: $csr:literal,
+     $mask:literal,
+     $($field_doc:expr, $field:ident: $bit:literal$(,)?)+) => {
+        $crate::read_only_csr! { $doc, $ty: $csr, $mask }
+
+        $(
+            $crate::write_only_csr_field! {
+                $ty,
+                $field_doc,
+                $field: $bit,
+            }
+        )+
+    };
+
+    ($doc:expr,
+     $ty:ident: $csr:literal,
+     $mask:literal,
+     $(
+     $field_doc:expr,
+     $field:ident: $bit_start:literal ..= $bit_end:literal$(,)?
+     )+) => {
+        $crate::read_only_csr! { $doc, $ty: $csr, $mask }
+
+        $(
+            $crate::write_only_csr_field! {
+                $ty,
+                $field_doc,
+                $field: $bit_start..=$bit_end,
+            }
+        )+
+
+    };
+
+    ($doc:expr,
+     $ty:ident: $csr:literal,
+     $mask:literal,
+     $(
+     $field_doc:expr,
+     $field:ident: [$bit_start:literal : $bit_end:literal]$(,)?
+     )+) => {
+        $crate::read_only_csr! { $doc, $ty: $csr, $mask }
+
+        $(
+            $crate::write_only_csr_field! {
+                $ty,
+                $field_doc,
+                $field: [$bit_start:$bit_end],
+            }
+        )+
+    };
+
+
+    ($doc:expr, $ty:ident: $csr:literal, $mask:literal) => {
+        #[repr(C)]
+        #[doc = $doc]
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub struct $ty {
+            bits: usize,
+        }
+
+        impl $ty {
+            /// Bitmask for legal bitfields of the CSR type.
+            pub const BITMASK: usize = $mask;
+
+            /// Creates a new CSR type from raw bits value.
+            ///
+            /// Only bits in the [BITMASK](Self::BITMASK) will be set.
+            pub const fn from_bits(bits: usize) -> Self {
+                Self { bits: bits & $mask }
+            }
+
+            /// Gets the raw bits value.
+            pub const fn bits(&self) -> usize {
+                self.bits & $mask
+            }
+
+            /// Gets the bitmask for the CSR type's bitfields.
+            pub const fn bitmask(&self) -> usize {
```

Comment:
Any reason not to prune this and just use `Self::BITMASK`?

Reply:
The main reason I included it is when you have:

```rust
let csr = SomeCsrType::from_bits(0);
let what_is_the_bitmask_again = csr.bitmask();
```

Or in situations where referring to the specific type isn't important. I guess this wouldn't be all that relevant unless we implemented a trait for CSR types.

For example, a collection of CSR types where the concrete types differ, but the user still wants to access the bitmask of each item. In that case `CsrType::BITMASK` is less ergonomic than `csr_var.bitmask`.

Reply:
The main reason I included it is when you have:

```rust
let csr = SomeCsrType::from_bits(0);
let what_is_the_bitmask_again = csr.bitmask();
```

Or in situations where referring to the specific type isn't important. I guess this wouldn't be all that relevant unless we implemented a trait for CSR types.

For example, a collection of CSR types where the concrete types differ, but the user still wants to access the bitmask of each item. In that case, `CsrType::BITMASK` is less ergonomic than `csr_var.bitmask()`.